### ISSUE TITLE
Contextual rules do not honor the agent source

### ISSUE THREAD

**Issue by crolopez** [wazuh/wazuh#3865]
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 3.9.5 | Analysisd | Manager | Packages and sources | Any |

Contextual rules, which are activated with a tag like `<same_source_ip>`, should check the ID of the agent (or manager) that generated the event. Since this does not happen, the trigger value (given by `frequency`) can be reached with the events generated by several agents.

For example, the rule `5712` is triggered if `5710` matches 8 times in 120 seconds.

``` XML
  <rule id="5712" level="10" frequency="8" timeframe="120" ignore="60">
    <if_matched_sid>5710</if_matched_sid>
    <description>sshd: brute force trying to get access to </description>
    <description>the system.</description>
    <same_source_ip />
    <group>authentication_failures,pci_dss_11.4,pci_dss_10.2.4,pci_dss_10.2.5,gdpr_IV_35.7.d,gdpr_IV_32.2,hipaa_164.312.b,nist_800_53_SI.4,nist_800_53_AU.14,nist_800_53_AC.7,</group>
  </rule>
```

This means that we are considering a brute force attack if someone tries to access an agent 8 times (by ssh) in that time without success. However, we can see this alert if 8 of our agents are accessed unsuccessfully once in that time period.

This is a historical issue in Wazuh, but I think that its solution is not complex.

We could verify that the event we are iterating comes from the same ID, ignoring it if not. It would be good to allow the evaluation of the timeframe to get out of the loop as soon as possible. For this reason, I think that the check could be made here:

- https://github.com/wazuh/wazuh/blob/6f47eb206f65ecef6053ff77921639161e1a8db6/src/analysisd/eventinfo.c#L78
- https://github.com/wazuh/wazuh/blob/6f47eb206f65ecef6053ff77921639161e1a8db6/src/analysisd/eventinfo.c#L298
- https://github.com/wazuh/wazuh/blob/6f47eb206f65ecef6053ff77921639161e1a8db6/src/analysisd/eventinfo.c#L507 (Remove `else if` in this case)

``` C
if (strcmp(lf->agent_id, my_lf->agent_id)) {
    continue; // Or goto next_it; in Search_LastEvents
}
```

In addition, with this feature, we would gain efficiency if in large environments.

If there is a rule where this behaviour is expected and it is desired that all agents increase the counter, we could add some option to indicate it.

``` XML
  <rule id="5712" level="10" frequency="8" timeframe="120" ignore="60">
    <if_matched_sid>5710</if_matched_sid>
    <description>sshd: brute force trying to get access to </description>
    <description>the system.</description>
    <count_all_agent />
    <same_source_ip />
    <group>authentication_failures,pci_dss_11.4,pci_dss_10.2.4,pci_dss_10.2.5,gdpr_IV_35.7.d,gdpr_IV_32.2,hipaa_164.312.b,nist_800_53_SI.4,nist_800_53_AU.14,nist_800_53_AC.7,</group>
  </rule>
```

**Comment by crolopez** [comment]
The current state of the solution implemented in (https://github.com/wazuh/wazuh/issues/3931) solves the problem by default, taking into account only the events whose agent ID matches. However, there is an option that evaluates [the location field](https://documentation.wazuh.com/3.9/user-manual/ruleset/ruleset-xml-syntax/rules.html?highlight=rule%20syntax#same-location), which is a key formed by the name of the agent, its ID and the origin of the event. 

We have to decide on the following:
- If we think that the `<same_location>` tag is enough, we will have to add this option to all the conceptual rules (223 currently). We will also need to extend its documentation.
- If we think that the `<same_location>` tag is not enough, either because users don't usually associate the location with the agent identity and we need an explicit way to name it, or because including the origin of the event in this primary key can generate invalid behaviours:
  - We can add the agent ID check by default and allow the user to block this check by another option like `<global_frequency>`, where the word global means that the engine will take into account the global counter of all agents to evaluate the alert triggering.
  - We can allow the check only if an explicit option specifies it, such as `<same_agent_id>`. This option would make us modify all the contextual rules.
  - We can also add an attribute to the `frequency` tag level that specifies it. This option would also make us modify all the implicated rules.
``` XML
  <rule id="11306" level="10" frequency="8" timeframe="3600" global="yes">
```

**Comment by chemamartinez** [comment]
Merged #3931
